src/rx-query.ts

```diff
@@ -490,6 +486,118 @@ export class RxQueryBase<
+    }
+
+    private async _restoreQueryCacheFromPersistedState() {
+        if (!this._queryCacheBackend) {
+            // no cache backend provided, do nothing
+            return;
+        }
+
+        if (this._persistentQueryCacheResult) {
+            // we already restored the cache once, no need to run twice
+            return;
+        }
+
+        if (this.mangoQuery.skip) {
+            console.error('The persistent query cache only works on non-skip queries.');
+            return;
+        }
+
+        const persistentQueryId = this.persistentQueryId();
+        const value = await this._queryCacheBackend.getItem<string[] | string>(`qc:${persistentQueryId}`);
+
+        const lwt = (await this._queryCacheBackend.getItem(`qc:${persistentQueryId}:lwt`)) as string ?? RX_META_LWT_MINIMUM;
+
+        const primaryPath = this.collection.schema.primaryPath;
+
+        this._persistentQueryCacheResult = value ?? undefined;
+
+        // if this is a regular query, also load documents into cache
+        if (Array.isArray(value) && value.length > 0) {
+            this._persistentQueryCacheResult = value;
+            this._persistentQueryCacheResultLwt = lwt;
+
+            const persistedQueryCacheIds = new Set(this._persistentQueryCacheResult);
+
+            let docsData: RxDocumentData<RxDocType>[] = [];
+
+            // query all docs updated > last persisted, limit to an arbitrary 1_000_000 (10x of what we consider our largest library)
+            const {documents: changedDocs} = await this.collection.storageInstance.getChangedDocumentsSince(
+              1_000_000,
+              // make sure we remove the monotonic clock (xxx.01, xxx.02) from the lwt timestamp to avoid issues with
+              // lookups in indices (dexie)
+              {id: '', lwt: Math.floor(Number(lwt))}
+            );
+
+            for (const changedDoc of changedDocs) {
+              /*
+               * no need to fetch again, we already got the doc from the list of changed docs, and therefore we filter
+               * deleted docs as well
+               */
+              persistedQueryCacheIds.delete(changedDoc[primaryPath] as string);
+
+              // ignore deleted docs or docs that do not match the query
+              if (!this.doesDocumentDataMatch(changedDoc)) {
+                continue;
+              }
+
+              // add to document cache
+              this.collection._docCache.getCachedRxDocument(changedDoc);
+
+              // add to docs
+              docsData.push(changedDoc);
+            }
+
+            // fetch remaining persisted doc ids
+            const nonRestoredDocIds = Array.from(persistedQueryCacheIds).filter(docId => {
+              // first try to fill from docCache
+              const docData = this.collection._docCache.getLatestDocumentDataIfExists(docId);
+              if (docData && this.doesDocumentDataMatch(docData)) {
+                docsData.push(docData);
+              }
```

Comment:
it's a bit strange to modify some captured variable `docsData` here inside of a `filter()` closure. normally those are separate

Reply:
This is doing the same, I am good with either.

```ts
const nonRestoredDocIds: string[] = [];
for (const docId of persistedQueryCacheIds) {
  // first try to fill from docCache
  const docData = this.collection._docCache.getLatestDocumentDataIfExists(docId);
  if (docData && this.doesDocumentDataMatch(docData)) {
    docsData.push(docData);
  }

  if (!docData) {
    nonRestoredDocIds.push(docId);
  }
}
```

Reply:
that definitely reads nicer to me if i ever had to understand & change this code.